### ISSUE TITLE
Eager loading doesn't work in some cases

### ISSUE THREAD

**Issue by nvloff** [mongoid/mongoid#3942]
Mongoid: 4 ( master at 35aea60 )

Having a criteria like this:

```
posts =  Post.all.includes(:person)
```

The first time the `posts` are iterated the person is not fetched from the db - but the second time it is.

```ruby
require 'mongoid'
require 'mongoid/support/query_counter'
require 'minitest/autorun'
# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

Mongoid.configure.connect_to("mongoid_test")

Mongoid.logger = Logger.new($stdout)
Moped.logger = Logger.new($stdout)

def count_queries(&block)
  query_counter = Mongoid::QueryCounter.new
  query_counter.instrument(&block)
  query_counter.events.size
end

class Post
  include Mongoid::Document
  belongs_to :person
end


class Person
  include Mongoid::Document
  has_many :posts
  field :name
end

class BugTest < Minitest::Test
  def test_query_count
    p = Person.create!(name: 'arthurnn')
    4.times { Post.create!(person: p) }

    criteria =  Post.all.includes(:person)

    query_counter = count_queries do
      criteria.each(&:person)
    end

    # this is fine
    assert_equal 2, query_counter

    query_counter = count_queries do
      criteria.each(&:person)
    end

    # this fails
    assert_equal 2, query_counter
  end
end
```

```
Finished tests in 0.114663s, 8.7212 tests/s, 17.4424 assertions/s.

  1) Failure:
BugTest#test_query_count [examples/mongoid_test.rb:49]:
Expected: 2
  Actual: 43

1 tests, 2 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by brauliomartinezlm** [comment]
Hi guys, so digging in a little bit, I found out what's going on. I just don't know what's the best way to fix it though :).

When the first each of your test executes, it eager loads person correctly because [this](https://github.com/mongoid/mongoid/blob/master/lib/mongoid/contextual/mongo.rb#L634) returns true as it should. Now the second time it runs, and as they share the same [execution context](https://github.com/mongoid/mongoid/blob/master/lib/mongoid/contextual.rb#L35), it returns false, so the actual [query](https://github.com/mongoid/mongoid/blob/master/lib/mongoid/contextual/mongo.rb#L639) gets executed.

Why it returns false the second time? Because [eager_loadable?](https://github.com/mongoid/mongoid/blob/master/lib/mongoid/relations/eager.rb#L25) depends on the value of [eager_loaded](https://github.com/mongoid/mongoid/blob/master/lib/mongoid/relations/eager.rb#L12) accessor which [gets set to true](https://github.com/mongoid/mongoid/blob/master/lib/mongoid/relations/eager.rb#L33) after the first time it executes for the first each.

So, I really don't know if the real problem is both each share the execution context of the query, if that accessor needs to be set back at some point or the actual use of it needs to be refactored... Any thoughts?

**Comment by p-run** [comment]
Is there a patch we can try?

**Comment by nvloff** [comment]
@p-run If you add `.cache` to your criteria - eager loading would work in *most* cases.

https://github.com/mongoid/mongoid/blob/master/lib/mongoid/contextual/mongo.rb#L631-L641

Basically here you don't want to enter the cached case and not do the query.

**Comment by brauliomartinezlm** [comment]
Yes, @nvloff is right, that's a possible workaround in app code. Anyways, we need to fix the actual issue in the gem. @p-run I have not started coding towards any path. I was waiting for some conceptual review on what way people thought it was to take.
